File: docker-cam-wrapper/camera_subscriber_cpp/src/camera_sub.cpp
#include <rclcpp/rclcpp.hpp>
#include <sensor_msgs/msg/image.hpp>
#include <cv_bridge/cv_bridge.h>
#include <opencv2/opencv.hpp>

class CameraSubscriber : public rclcpp::Node
{
public:
  CameraSubscriber()
  : Node("camera_subscriber")
  {
    sub_ = this->create_subscription<sensor_msgs::msg::Image>(
      "/image_raw",
      rclcpp::SensorDataQoS(),
      std::bind(&CameraSubscriber::image_callback, this, std::placeholders::_1));

    RCLCPP_INFO(this->get_logger(), "Camera subscriber started, waiting for images...");
  }

private:
  void image_callback(const sensor_msgs::msg::Image::SharedPtr msg)
  {
    try {
      cv::Mat frame = cv_bridge::toCvCopy(msg, "bgr8")->image;
      RCLCPP_INFO(this->get_logger(),
        "Received image: %dx%d", frame.cols, frame.rows);
    } catch (const cv_bridge::Exception & e) {
      RCLCPP_ERROR(this->get_logger(), "cv_bridge exception: %s", e.what());
    }
  }

  rclcpp::Subscription<sensor_msgs::msg::Image>::SharedPtr sub_;
};

int main(int argc, char ** argv)
{
  rclcpp::init(argc, argv);
  rclcpp::spin(std::make_shared<CameraSubscriber>());
  rclcpp::shutdown();
  return 0;
}

File: docker-cam-wrapper/camera_subscriber/camera_subscriber/camera_sub.py
import rclpy
from rclpy.node import Node
from rclpy.qos import qos_profile_sensor_data
from sensor_msgs.msg import Image
from cv_bridge import CvBridge, CvBridgeError


class CameraSubscriber(Node):
    def __init__(self):
        super().__init__('camera_subscriber')

        self.bridge = CvBridge()
        self.frame_count = 0

        self.subscription = self.create_subscription(
            Image,
            '/image_raw',
            self.image_callback,
            qos_profile_sensor_data
        )

        self.get_logger().info('Camera subscriber started, waiting for images on /image_raw ...')

    def image_callback(self, msg: Image):
        self.frame_count += 1

        try:
            frame = self.bridge.imgmsg_to_cv2(msg, desired_encoding='bgr8')

            self.get_logger().info(
                f'[RECV OK] frame#{self.frame_count} '
                f'size={frame.shape[1]}x{frame.shape[0]} '
                f'encoding={msg.encoding} '
                f'stamp={msg.header.stamp.sec}.{msg.header.stamp.nanosec:09d}'
            )

        except CvBridgeError as e:
            self.get_logger().error(
                f'[RECV ERROR] frame#{self.frame_count} CvBridgeError: {e}'
            )


def main(args=None):
    rclpy.init(args=args)
    node = CameraSubscriber()

    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        pass
    finally:
        node.destroy_node()
        rclpy.shutdown()


if __name__ == '__main__':
    main()
